ProductService_and_tests

AntonBelousov2020 opened the pull request:
Add methods calculateStock, receiveDelivery, reserveProduct on ProductServiceImpl class, add tests for this methods on ProductServiceImplTest class

Raccoon2304 committed:
Add methods calculateStock, receiveDelivery, reserveProduct on ProductServiceImpl class, add tests for this methods on ProductServiceImplTest class

MikhailBelyaev03 commented on store-impl/src/main/java/com/epam/rd/service/impl/ProductServiceImpl.java:
```diff
@@ -1,23 +1,137 @@
+
+        List<Product> productList = productRepository.findAll();
+
+        productMap = productList.stream().filter(product -> this.getQuantity(product) < 100)
```
Каждый новый метод стрима лучше на новой строчке делать

AntonBelousov2020 replied:
Поправил

MikhailBelyaev03 commented on store-impl/src/main/java/com/epam/rd/service/impl/ProductServiceImpl.java:
```diff
@@ -1,23 +1,137 @@
+                        this::getNeedQuantity));
 
+        if (productMap.isEmpty()) {
+            log.warn("Quantity for all products more then 100");
```
Это не совсем правильно. Если все у всех товаров количество больше 100, то это не ошибка. Поэтому ничего не надо выбрасывать. Просто не создаешь заказ, для доп инфы можешь уровнем инфо написать, что все товары в наличии.

AntonBelousov2020 replied:
Поправил

MikhailBelyaev03 commented on store-impl/src/main/java/com/epam/rd/service/impl/ProductServiceImpl.java:
```diff
@@ -1,23 +1,137 @@
 
+        supplierOrder.getSupplierOrderItems().forEach(supplierOrderItem -> {
+            Product product = supplierOrderItem.getProduct();
+            product.getCatalog().setQuantity(calculateNewQuantity(product, supplierOrderItem));
```
А сохранить товары забыл?

AntonBelousov2020 replied:
Добавил

MikhailBelyaev03 commented on store-impl/src/main/java/com/epam/rd/service/impl/ProductServiceImpl.java:
```diff
@@ -1,23 +1,137 @@
+        });
+
+        Catalog catalog = product.getCatalog();
+        catalog.setQuantity(this.getQuantity(product) - quantity);
```
давай лучше переименуем метод this.getQuantity в чтото типа extractQuantity и вызывать будем без this

AntonBelousov2020 replied:
Поправил

MikhailBelyaev03 commented on store-impl/src/main/java/com/epam/rd/service/impl/ProductServiceImpl.java:
```diff
@@ -1,23 +1,137 @@
+        });
+
+        Catalog catalog = product.getCatalog();
+        catalog.setQuantity(this.getQuantity(product) - quantity);
```
А где проверка, что количество товар может быть в наличии меньше, чем нужно зарезервировать?

AntonBelousov2020 replied:
Здесь правильнее будет кинуть рантайм эксепшон или просто вывести лог с уровнем info?

MikhailBelyaev03 replied:
Здесь уже ошибка должна вылетать, так что да RuntimeEx

AntonBelousov2020 replied:
Поправил

Raccoon2304 committed:
Move to new lines every methods from the stream, remove wrongs throw exception in method calculateStock, add save items to catalog in method receiveDelivery, rename method from getQuantity to extractQuantity, remove redundant "this." in method reserveProduct, add check quantity of goods in stock, add throw new RuntimeException in method reserveProduct on ProductServiceImpl class.
Remove redundant check Runtime Exception in method testCalculateStockWhenMapEmpty, add verify(catalogRepository).save(catalog) in method testReceiveDeliveryWhenDeliveredSupplierOrderIdExists, add externalQuantity of product for more correctly work method testReserveProductWhenProductExists, add externalQuantity of product for more correctly work method testReserveProductWhenProductNotExists, add new method testReserveProductWhenTheQuantityOfGoodsLessThenNeedForOrder, add new private method createSupplierOrderItemList for more correctly work method testReceiveDeliveryWhenDeliveredSupplierOrderIdExists on ProductServiceImplTest class
